spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -0,0 +1,217 @@
+import serialization.Bigdl.{AttrValue, BigDLModule}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * [[RecurrentDecoder]] module is a container of rnn cells that used to make
+ * a prediction of the next timestep based on the prediction we made from
+ * the previous timestep. Input for RecurrentDecoder is dynamically composed
+ * during training. input at t(i) is output at t(i-1), input at t(0) is
+ * user input, and user input has to be batch x ???(depends on cell type)
+ * without time information.
+
+ * Different types of rnn cells can be added using add() function. Currently
+ * only support lstmpeephole, convlstm, convlstm3D cell.
+ */
+class RecurrentDecoder[T : ClassTag](seqLength: Int)
+  (implicit ev: TensorNumeric[T]) extends Recurrent[T] {
+
+  times = seqLength
+  private val newInput = Tensor[T]()
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim == 2 || input.dim == 4 || input.dim == 5,
+      "Recurrent: input should be a 2D/4D/5D Tensor, e.g [batch, nDim], " +
+        s"current input.dim = ${input.dim}")
+
+    batchSize = input.size(batchDim)
+
+    val hiddenSize = topology.hiddensShape(0)
+    val outputSize = input.size()
+    outputSize(1) = hiddenSize
+    require(hiddenSize == input.size()(1), "hiddenSize is " +
+      "not the same with input size!! Please update cell settings or use Recurrent instead!")
+    val featureSizes = outputSize.drop(1)
+    output.resize(Array(batchSize, times) ++ featureSizes)
+    // Clone N modules along the sequence dimension.
+    extend(featureSizes)
+    if (preTopology != null) newInput.resize(output.size())
+    else outputCell.resize(output.size())
+
+    /**
+     * currentInput forms a T() type. It contains two elements, hidden and input.
+     * Each time it will feed the cell with T(hidden, input) (or T(input, hidden) depends on
+     * your hidDim and inputDim), and the cell will give a table output containing two
+     * identical elements T(output, output). One of the elements from the cell output is
+     * the updated hidden. Thus the currentInput will update its hidden element with this output.
+     */
+    var i = 1
+    // init state
+    currentInput(hidDim) = if (initState != null) initState
+     else hidden
+
+    while (i <= times) {
+      // input at t(0) is user input
+      val inputTmp = if (i == 1) {
+        input
+      } else {
+        // input at t(i) is output at t(i-1)
+        cells(i - 2).output.toTable[Tensor[T]](inputDim)
+      }
+
+      currentInput(inputDim) = if (preTopology != null) {
+        newInput.narrow(2, i, 1).copy(inputTmp)
+        val sizes = 1 +: inputTmp.size()
+        inputTmp.resize(sizes)
+        val _input = preTopology.updateOutput(inputTmp).toTensor[T]
+        inputTmp.resize(sizes.takeRight(sizes.length - 1))
+        _input.select(1, 1)
+      } else {
+        outputCell.narrow(2, i, 1).copy(inputTmp)
```

Comment:
no need to copy; just maintain an array of inputTmp, which can be then used in backward